Agents architecture page part 1 - done (#3511)
jenkins-infra/jenkins.io: content/doc/book/using/using-agents.adoc

---
layout: section
title: Using Jenkins agents
---
ifdef::backend-html5[]
:description:
:author:
:email: jenkinsci-docs@googlegroups.com
:sectanchors:
:toc:
:toclevels: 4
:hide-uri-scheme:
ifdef::env-github[:imagesdir: ../resources]
ifndef::env-github[:imagesdir: ../../resources]
endif::[]

> = Jenkins agents architecture
> 

Reviewer: This text is ignored.  The level 1 heading is inserted by the framework using the value of `title` from the prolog.
```suggestion
```

* SSH key pair

[NOTE]
====
If you need help to install Java, Jenkins and Docker please visit the section link:/doc/book/installing/docker/[Installing Jenkins.]
====

=== Generating an SSH key pair

1. In a terminal window run the command: `ssh-keygen -f ~/.ssh/jenkins_agent_key`
2. Provide a passphrase to use with the key
3. Confirm the output looks something like this:
+
[source,bash]
----
ubuntu@desktop:~$ ssh-keygen -f ~/.ssh/jenkins_agent_key
Generating public/private rsa key pair.
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in /home/ubuntu/.ssh/jenkins_agent_key
Your public key has been saved in /home/ubuntu/.ssh/jenkins_agent_key.pub
The key fingerprint is:
SHA256:XqxxjqsLlvDD0ZHm9Y2iR7zC6IbsUlMEHo3ffy8TzGs
The key's randomart image is:
+---[RSA 3072]----+
|  o+             |
| ...o  .         |
|  .o .+ .        |
|    o+.+ o o     |
|  ... o.So* .    |
|  o+ = +.X=      |
| o oO + *..+     |
|. oo.o o .E .    |
| o... oo.. o     |
+----[SHA256]-----+
----

==== Create a Jenkins SSH credential

1. Go to your Jenkins dashboard;
2. Go to `Manage Jenkins` option in main menu and click on credentials button;
+

> image:node/credentials-1.png[credentials menu]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
image:node/credentials-1.png[credentials menu,700]
```

3. select the drop option `Add Credentials` from the global item;
+

> image:node/credentials-2.png[add credentials option]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
image:node/credentials-2.png[add credentials option,700]
```

4. Fill the form:
* Kind:  SSH Username with private key;
* id: jenkins
* description: The jenkins ssh key
* username: jenkins
* Private Key: select `Enter directly` and press the Add button to insert your private key from `~/.ssh/jenkins_agent_key`
* Passphrase: fill your passphrase used to generate the SSH key pair and then press OK

>     image:node/credentials-3.png[credentials filled form]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
    image:node/credentials-3.png[credentials filled form,700]
```

* Passphrase: fill your passphrase used to generate the SSH key pair and then press OK
    image:node/credentials-3.png[credentials filled form]

=== Creating your Docker agent

Here we will use the link:https://github.com/jenkinsci/docker-ssh-agent[docker-ssh-agent image] to create the agent containers.

1. run the command to start your first agent:
+
[source,bash]
----
docker run -d --rm --name=agent1 -p 22:22 \
-e "JENKINS_AGENT_SSH_PUBKEY=[your-public-key]" \
jenkins/ssh-agent:alpine
----
+
[NOTE]
====
* Remember to replace the tag [your-public-key] for your own SSH *public* key.
* Your public key in this example is: `cat ~/.ssh/jenkins_agent_key.pub`
====

> 2. Now run the follow command to update the container environment:

Reviewer: ```suggestion
2. Now run the following command to update the container environment:
```

----

> docker exec  agent1 sh -c "env | egrep -v '^(HOME=|USER=|MAIL=|LC_ALL=|LANG=|HOSTNAME=|TERM=|LANGUAGE=)' >> /etc/environment"

Reviewer: I should have read the command more carefully before reducing line length.  The `-v` argument inverts the sense of the `egrep` matching.
```suggestion
docker exec  agent1 sh -c "env | egrep -v "^(HOME=|USER=|MAIL=|LC_ALL=|LS_COLORS=|LANG=|HOSTNAME=|PWD=|TERM=|SHLVL=|LANGUAGE=|_=)" >> /etc/environment"
```

> [source,bash]
> ----
> docker exec  agent1 sh -c "env | egrep -v "^(HOME=|USER=|MAIL=|LC_ALL=|LS_COLORS=|LANG=|HOSTNAME=|PWD=|TERM=|SHLVL=|LANGUAGE=|_=)" >> /etc/environment"
> ----

Reviewer: Narrow the example source code so that it does not disrupt screen layout for the entire page.
```suggestion
[source,bash]
----
$ VARS1="HOME=|USER=|MAIL=|LC_ALL=|LS_COLORS=|LANG="
$ VARS2="HOSTNAME=|PWD=|TERM=|SHLVL=|LANGUAGE=|_="
$ VARS="${VARS1}|${VARS2}"
$ docker exec agent1 sh -c "env | egrep -v "^(${VARS})" >> /etc/environment"
----
```

> docker exec  agent1 sh -c "env | egrep -v '^(HOME=|USER=|MAIL=|LC_ALL=|LS_COLORS=|LANG=|HOSTNAME=|PWD=|TERM=|SHLVL=|LANGUAGE=|_=)' >> /etc/environment"

Reviewer: Anything we can do to narrow this line will help the entire page when rendered on narrower screens.  

The `LS_COLORS` variable controls color output of the `ls` command.  That seems optional in a Docker agent.

The `SHLVL`,  `PWD`, and `_` variables will be set by the shell.

```suggestion
docker exec  agent1 sh -c "env | egrep -v '^(HOME=|USER=|MAIL=|LC_ALL=|LANG=|HOSTNAME=|TERM=|LANGUAGE=)' >> /etc/environment"
```

----
+
[NOTE]
====
The step 2 is necessary because the image is configured to reject changes to environment variables.
when the link:https://github.com/jenkinsci/docker-ssh-agent/issues/33[issue #33] is fixed, we can ignore this step.
====

> 3. Now the container agent1 it is running. +

Reviewer: Remove extra word and highlight the name of the agent for easier reading.
```suggestion
3. Now the container `agent1` is running. +
```

=== Setup up the agent1 on jenkins.

1. Go to your Jenkins dashboard;
2. Go to `Manage Jenkins` option in main menu;
3. Go to `Manage Nodes and clouds` item;
+

> image:node/node-1.png[Menage node menu]

Reviewer: Spelling fix
```suggestion
image:node/node-1.png[Manage node menu]
```

> image:node/node-1.png[Manage node menu]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
image:node/node-1.png[Manage node menu,700]
```

4. Go to `New Node` option in side menu;
5. Fill the Node/agent name and select the type; (e.g. Name: agent1, Type: Permanent Agent)
6. Now fill the fields:
** Remote root directory; (e.g.: /home/jenkins )
** label; (e.g.: agent1 )
** usage; (e.g.: only build jobs with label expression...)
** Launch method; (e.g.: Launch agents by SSH )
*** Host; (e.g.: localhost or your IP address )
*** Credentials; (e.g.: jenkins )
*** Host Key verification Strategy; (e.g.: Manually trusted key verification ... )

>     image:node/node-2.png[node create form]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
    image:node/node-2.png[node create form,650]
```

7. Press the button save and the agent1 will be registered, but offline. Click on it.
+

> image:node/node-3.png[node offline]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
image:node/node-3.png[node offline,650]
```

8. Now press the button `Launch agent` and wait some seconds, then you should receive +
the message: `Agent successfully connected and online` on the last log line.
+

> image:node/node-4.png[Agent successfully connected]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
image:node/node-4.png[Agent successfully connected,650]
```

=== Delegating the first job to agent1

1. Go to your jenkins dashboard;
2. Select `New Item` on side menu;
3. Enter a name. (e.g.: First Job to Agent1)
4. Select the `Freestyle project` and press OK;
5. Check the option: `Restrict where this project can be run`;
6. Fill the field: label with the agent1 label; (e.g.: agent1)
+

> image:node/node-5.png[Agent job 1]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
image:node/node-5.png[Agent job 1,650]
```

+
[NOTE]
====
Be careful with white spaces before or after the label.
====

7. Now Select the option `Execute shell` at Build Section;
+

> image:node/node-6.png[Agent job 2]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
image:node/node-6.png[Agent job 2,650]
```

+
image:node/node-6.png[Agent job 2]

> 8. Fill the with command: `NODE_NAME` and the name +

Reviewer: Phrasing improvement.
```suggestion
8. Add the command: `echo $NODE_NAME` in the `Command` field of the `Execute shell` step  and the name +
```

> of the agent will be printed inside the log when this job run;

Reviewer: ```suggestion
of the agent will be printed inside the log when this job is run;
```

+

> image:node/node-7.png[Agent job 3]

Reviewer: Limit width of image, allow user to open full size image in another window or tab
```suggestion
image:node/node-7.png[Agent job 3,650]
```

+
image:node/node-7.png[Agent job 3]

11. you should receive output similar to:
+
[source,bash]
----
Started by user Admin User
Running as SYSTEM
Building remotely on agent1 in workspace /home/jenkins/workspace/First Job to Agent1

> [First Job to Agent1] $ /bin/sh -xe /tmp/jenkins15623311211559049312.sh
> Finished: SUCCESS

Reviewer: Needs the example text added in the annotated location
```suggestion
[First Job to Agent1] $ /bin/sh -xe /tmp/jenkins15623311211559049312.sh
$ echo $NODE_NAME
agent1
Finished: SUCCESS
```

Enter same passphrase again:
Your identification has been saved in /home/ubuntu/.ssh/jenkins_agent_key
Your public key has been saved in /home/ubuntu/.ssh/jenkins_agent_key.pub
The key fingerprint is:
SHA256:XqxxjqsLlvDD0ZHm9Y2iR7zC6IbsUlMEHo3ffy8TzGs
The key's randomart image is:
+---[RSA 3072]----+
|  o+             |
| ...o  .         |
|  .o .+ .        |
|    o+.+ o o     |
|  ... o.So* .    |
|  o+ = +.X=      |
| o oO + *..+     |
|. oo.o o .E .    |
| o... oo.. o     |
+----[SHA256]-----+
----

==== Create a Jenkins SSH credential

1. Go to your Jenkins dashboard;
2. Go to `Manage Jenkins` option in main menu and click on credentials button;
+
image:node/credentials-1.png[credentials menu,700]

3. select the drop option `Add Credentials` from the global item;
+
image:node/credentials-2.png[add credentials option,700]

4. Fill the form:
* Kind:  SSH Username with private key;
* id: jenkins
* description: The jenkins ssh key
* username: jenkins
* Private Key: select `Enter directly` and press the Add button to insert your private key from `~/.ssh/jenkins_agent_key`
* Passphrase: fill your passphrase used to generate the SSH key pair and then press OK

>     image:node/credentials-3.png[credentials filled form,700]

Reviewer: ```suggestion
    image:node/credentials-3.png[credentials filled form,650]
```

$ VARS="${VARS1}|${VARS2}"
$ docker exec agent1 sh -c "env | egrep -v "^(${VARS})" >> /etc/environment"
----
+
[NOTE]
====
The step 2 is necessary because the image is configured to reject changes to environment variables.
when the link:https://github.com/jenkinsci/docker-ssh-agent/issues/33[issue #33] is fixed, we can ignore this step.
====
3. Now the container `agent1` is running. +
Hint: the command `docker ps` can be used to check if the container is running as expected.

=== Setup up the agent1 on jenkins.

1. Go to your Jenkins dashboard;
2. Go to `Manage Jenkins` option in main menu;
3. Go to `Manage Nodes and clouds` item;
+
image:node/node-1.png[Manage node menu,700]

4. Go to `New Node` option in side menu;
5. Fill the Node/agent name and select the type; (e.g. Name: agent1, Type: Permanent Agent)
6. Now fill the fields:
** Remote root directory; (e.g.: /home/jenkins )
** label; (e.g.: agent1 )
** usage; (e.g.: only build jobs with label expression...)
** Launch method; (e.g.: Launch agents by SSH )
*** Host; (e.g.: localhost or your IP address )
*** Credentials; (e.g.: jenkins )
*** Host Key verification Strategy; (e.g.: Manually trusted key verification ... )

>     image:node/node-2.png[node create form,650]

Reviewer: Deeply nested image needs to be smaller
```suggestion
    image:node/node-2.png[node create form,600]
```

* Jenkins installation
* Docker installation
* SSH key pair

[NOTE]
====
If you need help to install Java, Jenkins and Docker please visit the section link:https://www.jenkins.io/doc/book/installing/#installation-platforms[Installing Jenkins.]
====

=== Generating an SSH key pair

1. In a terminal window run the command: `ssh-keygen -f ~/.ssh/jenkins_agent_key`
2. Provide a passphrase to use with the key
3. Confirm the output looks something like this:
+
[source,bash]
----
ubuntu@desktop:~$ ssh-keygen -f ~/.ssh/jenkins_agent_key
Generating public/private rsa key pair.
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in /home/ubuntu/.ssh/jenkins_agent_key
Your public key has been saved in /home/ubuntu/.ssh/jenkins_agent_key.pub
The key fingerprint is:
SHA256:XqxxjqsLlvDD0ZHm9Y2iR7zC6IbsUlMEHo3ffy8TzGs
The key's randomart image is:
+---[RSA 3072]----+
|  o+             |
| ...o  .         |
|  .o .+ .        |
|    o+.+ o o     |
|  ... o.So* .    |
|  o+ = +.X=      |
| o oO + *..+     |
|. oo.o o .E .    |
| o... oo.. o     |
+----[SHA256]-----+
----

==== Create a Jenkins SSH credential

1. Go to your Jenkins dashboard;

> 2. Go to `Manage Jenkins` option in main menu;

Reviewer 2: Is this missing a step to click on the credentials button?

+
image:node/credentials-1.png[credentials menu]

> 3. select the drop option `add Credentials` from the global item;

Reviewer 2: All other steps are capitalised

Author: Information updated. Thank you!

4. Fill the form:
* Kind:  SSH Username with private key;
* id: jenkins
* description: The jenkins ssh key
* username: jenkins
* Private Key: select `Enter directly` and press the Add button to insert your private key from `~/.ssh/jenkins_agent_key`
* Passphrase: fill your passphrase used to generate the SSH key pair and then press OK
    image:node/credentials-3.png[credentials filled form]

=== Creating your Docker agent

Here we will use the link:https://github.com/jenkinsci/docker-ssh-agent[docker-ssh-agent image] to create the agent containers.

1. run the command to start your first agent:
+
[source,bash]
----
docker run -d --rm --name=agent1 -p 22:22 \
-e "JENKINS_AGENT_SSH_PUBKEY=[your-public-key]" \
jenkins/ssh-agent:alpine
----
+
[NOTE]
====
* Remember to replace the tag [your-public-key] for your own SSH *public* key.
* Your public key in this example is: `cat ~/.ssh/jenkins_agent_key.pub`
====
2. Now run the follow command to update the container environment:
+
[source,bash]
----
docker exec  agent1 sh -c "env | egrep -v '^(HOME=|USER=|MAIL=|LC_ALL=|LS_COLORS=|LANG=|HOSTNAME=|PWD=|TERM=|SHLVL=|LANGUAGE=|_=)' >> /etc/environment"
----
+
[NOTE]
====

> The step 2 is necessary because the image is configured to reject changes to environment variables.
> when the link:https://github.com/jenkinsci/docker-ssh-agent/issues/33[issue #33] fix it so we can ignore this step.

Reviewer: I would very much prefer to see the issue fixed in the Docker ssh agent image rather than having this included in the documentation.  It is complicated and difficult to understand.

Alternately, we could tell the users to open the "Advanced" section and enter the Java path from that location.
```suggestion
The step 2 is necessary because the image is configured to reject changes to environment variables.
when the link:https://github.com/jenkinsci/docker-ssh-agent/issues/33[issue #33] fix it so we can ignore this step.
```

Author: You are right Mark, but unfortunately, this kind of 'problems' must be expected for docker users. In fact, I wasted hours investigating and studying why the image didn't work as expected.

That command was the better workaround I handled to solve the situation and the user can add it to their CI pipeline script to execute agents on demand. 

For all those reasons I vote to keep that information inside the guide. 

So, what should I do?

====

> 3. Now the container agent1 it is running and ready to connections. +

Reviewer 2: Remove it, ready for connections

Author: Information removed. Thank you!

execute the command `docker ps` to check if the container is running, something similar will be shown:
+
[source,bash]
----
ubuntu@desktop:~$ docker ps
CONTAINER ID        IMAGE                     COMMAND                  CREATED             STATUS              PORTS                    NAMES
7bf0214a9bfa        jenkins/ssh-agent:jdk11   "setup-sshd"             31 seconds ago      Up 28 seconds       0.0.0.0:22->22/tcp       agent1
d0116c8021c0        ruby:2.6                  "bundle exec awestru…"   41 minutes ago      Up 41 minutes       0.0.0.0:4242->4242/tcp   eloquent_lamport
ubuntu@desktop:~$
----

> === Setup up the agent1 on jenkins.

Reviewer 2: Remove up

Author: Information removed. Thank you!

> 1. Go to your jenkins dashboard;

Reviewer 2: Capitalise Jenkins

Author: Information updated. Thank you!

7. Press the button save and the agent1 will be registered, but offline. Click on it.
+
image:node/node-3.png[node offline]
8. Now press the button `Launch agent` and wait some seconds, then you should receive +
the message: `Agent successfully connected and online` on the last log line.
+
image:node/node-4.png[Agent successfully connected]

=== Delegating the first job to agent1

1. Go to your jenkins dashboard;
2. Select `New Item` on side menu;
3. Enter a name. (e.g.: First Job to Agent1)
4. Select the `Freestyle project` and press OK;
5. Check the option: `Restrict where this project can be run`;
6. Fill the field: label with the agent1 label; (e.g.: agent1)
+
image:node/node-5.png[Agent job 1]

+
[NOTE]
====
Be careful with white spaces before or after the label.
====

7. Now Select the option `Execute shell` at Build Section;
+
image:node/node-6.png[Agent job 2]

8. Fill the with command: `NODE_NAME` and the name +
of the agent will be printed inside the log when this job run;
9. press the save button and then select the option `Build Now`;
10. Wait some seconds ant the go to `Console Output` page
+
image:node/node-7.png[Agent job 3]

> 11. you should receive a output similar to:

Reviewer 2: Remove a

Author: Sorry , I didn't get it? What needs to be removed?

Author: Ho. I get it now, remove the letter 'a'. Ok. it was removed. Thank you
